### PR TITLE
Replace LevelDB with Bolt

### PR DESCRIPTION
**Pull request by rawfalafel**
### Important changes

- `db.GetX` will return a nil object instead of returning an error. i.e. `block, err := db.GetBlock(hash)` will return `nil` for both `block` and `err`. 
- `db.HasX` will no longer return an error, just the bool :)
- `db.NewDB` no longer automatically initializes genesis state. `InitializeState` needs to be explicitly called after instantiating the database. IMO, coupling the constructor with state changes makes the database harder to test.
- `ActiveState` and `CrystallizedState` are no longer cached in the db object. I'll explain why further down.

### How to use Bolt
The code becomes slightly more complex for simple operations, but there's a good reason for that. Most operations use a callback function. 

Here's `GetBlock`:
```go
func (db *BeaconDB) GetBlock(hash [32]byte) (*types.Block, error) {
	var block *types.Block
	err := db.view(func(tx *bolt.Tx) error {
		b := tx.Bucket(blockBucket)

		enc := b.Get(hash[:])
		if enc == nil {
			return nil
		}

		var err error
		block, err = createBlock(enc)
		return err
	})

	return block, err
}
```
The reason there's a callback function is because operations are handled as an atomic transaction, so there's extra code being called before/after the callback by the library. See `db.UpdateChainHead` for an example where multiple keys are updated atomically. Overall this pushes complexity down from the service layer to the database layer, which is a Good Thing™.

A "bucket" is a grouping of keys. You can treat them like key prefixes, which means that the same key in two buckets won't collide. So far I created `attestationBucket`, `blockBucket`, `mainChainBucket`, and `chainInfoBucket`. These more or less eliminate the need for key prefixes. 

### Why are we not caching state anymore?
The current database design needs to be changed to implement the IMD fork choice rule, and that includes changing the way state is stored and cached. 

The two main issues that I see:
- We need the ability to retrieve multiple `ActiveState` and `CrystallizedState` objects. This is important for handling cases where we receive a block from an older slot, or we need to run the fork choice rule from a previous cycle. Of course we don't need to store every state from genesis, but we need to strike a reasonable balance between space efficiency and speed.
- A more efficient way to handle reorgs. Currently, we assume that each new block builds cleanly on top of the current canonical chain. We store the blockhash for each slot in the canonical chain, and we need a better way to update this for larger reorgs.

If this PR looks good, I can create tickets for the two issues above.

### CODE REVIEW

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -147,10 +147,12 @@ func (c *ChainService) updateHead(processedBlock <-chan *types.Block) {
 				continue
 			}
 
-			cState := c.beaconDB.GetCrystallizedState()
-			aState := c.beaconDB.GetActiveState()
-			var stateTransitioned bool
+			aState, cState, err := c.beaconDB.GetState()
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go]: get state can sound a bit ambiguous - am i retrieving the "canonical" state or is it a generic function to retrieve any state pair from the db even from previous blocks?

File: beacon-chain/db/db.go [hunk 2]
```diff
@@ -1,146 +1,60 @@
...
-		state: &beaconState{},
-	}
-	hasCrystallized, err := beaconDB.has(crystallizedStateLookupKey)
-	if err != nil {
+func NewDB(dirPath string) (*BeaconDB, error) {
+	if err := os.MkdirAll(dirPath, 0700); err != nil {
 		return nil, err
 	}
-	hasGenesis, err := beaconDB.HasCanonicalBlockForSlot(0)
+	datafile := path.Join(dirPath, "rbc.db")
```
Comment (rauljordan) [inline beacon-chain/db/db.go]: maybe rename to something that isn't an acronym, like beaconchain.db

**Comment by rauljordan** [comment]
PR is excellent and quite easy to follow, thank you. For the items mentioned above:

> We need the ability to retrieve multiple ActiveState and CrystallizedState objects. This is important for handling cases where we receive a block from an older slot, or we need to run the fork choice rule from a previous cycle. Of course we don't need to store every state from genesis, but we need to strike a reasonable balance between space efficiency and speed.
A more efficient way to handle reorgs. Currently, we assume that each new block builds cleanly on top of the current canonical chain. We store the blockhash for each slot in the canonical chain, and we need a better way to update this for larger reorgs.

do you mean having some sort of batch get functionality for states in a range of slots or so? Should we not be storing the blockhash for each processed block in the db? Would be good to definitely open up some issues related to efficiency in reorgs, as that's the natural thing to work on after #643.

**Comment by rawfalafel** [comment]
> do you mean having some sort of batch get functionality for states in a range of slots or so?

Ah, my wording wasn't clear. I meant that we should be able to retrieve more than just the current head's active and crystallized state. Specifically, when we're running the fork choice rule and trying to decide between two blocks, we should be able to fetch the active and crystallized state for both blocks. Normally, we can expect the state to be in memory inside a map structure. However, when `beacon-node` restarts, the state for these blocks should be fetched from the database.

File: beacon-chain/blockchain/service.go (hunk 1 above)
Comment (rawfalafel) [reply beacon-chain/blockchain/service.go]: good point. i'm going to remove this, as I think it's going to be deprecated soon anyways.

**Pull request merged**